remote.go

```diff
@@ -1104,6 +1122,31 @@ func pushHashes(
 	return rs, nil
 }
 
+func (r *Remote) pruneShallow() error {
+	existingShallows, err := r.s.Shallow()
+	if err != nil {
+		return err
+	}
+
+	shallows := []plumbing.Hash{}
+
+	for _, shallow := range existingShallows {
+		if c, err := object.GetCommit(r.s, shallow); err == nil {
+			anyParents := false
+			for i := 0; i < c.NumParents(); i++ {
+				p, _ := c.Parent(i)
+				anyParents = anyParents || p != nil
+			}
+
+			if !anyParents {
+				shallows = append(shallows, shallow)
+			}
```

Comment:
Working with the assumption that a commit is only considered to be shallow if _none_ of its parents exist locally.

Reply:
This assumption might bite us – wouldn't it fail to be true if we loaded two files untracked from simultaneous commits?

Reply:
Yeah this is just the wrong approach. I don't think it's necessarily to do with that assumption, but the intention from the protocol is that the server is giving the client some explicit directions as to which commits will be shallow/unshallow as a result of processing the packfile. I just pushed an update to use the protocol instead of manually pruning.